README.md
# log-credentials
PAM log-credentials has been written by José Miguel Silva Caldeira <miguel@ncdc.pt>.

## Description:
PAM log-credentials allows you to log credentials.

It will log messages like this:
* May 22 18:50:15 local host log credentials: host = xxx.xxx.xxx.xxx service = sshd user = root root = xxxxxxx ** in system syslog;
* 1573749689 host = 192.168.1.223 service = sshd user = aaaaaaaaaaaaaaaa pass = aaaaaaaaaaaaaaa in the file if configured.

The idea of this module is to facilitate obtaining information that allows you to analyze brute force attacks on ssh and other services.

Through these data, with scripts and in a simple way it is possible to perceive:
* From where the attacks are initiated, the countries, the networks, etc. Through IP;
* You can see which users are most used in login attempts;
* The type of passwords that are used in attack attempts;
* Among others.

In this way it is possible for the administrator to be aware and to take measures so that good practices are taken to avoid these attacks.

Use at your own risk!

Do not use on a production system where multiple users have access to the logs. If you do, everyone has access to all credentials.

If you still want to do so, configure the log file access permissions.

## Tested Services / Commands
The module has been tested with the services / commands:
* sshd service;
* sudo command;
* su command.

## Ethic
Since this module logs all the credentials it is clear that it also logs the correct ones.

This module can be used as "KeyLogger" if it is badly installed on some system.

My intention is to use it for study purposes but stay of the conscience of each one uses it for a certain purpose.

Be cool!

## Installation

### Clone the project
You can clone the repository wherever you want. (I like to keep it in `~/Projects/log-credentials`.)

```Bash
$ cd ~/Projects
$ git clone https://github.com/jomisica/log-credentials.git
```

### Dependent software
In order to compile this software it is necessary to first install the dependencies. This will depend on the system in question below is the example for CentOS and Ubuntu.

#### CentOS 7/8
In CentOS it is necessary to install the "Development Tools" group that installs the necessary software to compile this software.

It is also necessary to install the package "pam-devel" that has the necessary headers so that it is possible to develop modules for the PAM in the system that we use.

```Bash
# yum group install "Development Tools"
# yum install pam-devel
```

#### Ubuntu 16.04
In Ubuntu it is necessary to install the software dependent to compile this software.

It is also necessary to install the package "libpam0g-dev" that has the necessary headers so that it is possible to develop modules for the PAM in the system that we use.

```bash
# apt-get build-dep pam
# apt-get install libpam0g-dev
```

### Build
```Bash
$ cd log-credentials
$ libtoolize --force
$ aclocal
$ autoheader
$ automake --force-missing --add-missing
$ autoconf
```

The directory where the PAM and NSS modules are installed are different on several systems, as such, the correct location may be required when configuring.

```bash
$ ./configure
...
...
...
...
configure: Summary of build options:
  Version:              1.0.1
  Host type:            x86_64-unknown-linux-gnu
  Compiler:             gcc
  CFLAGS:               -g -O2
  Library types:        Shared=yes, Static=no
  PAM Module Directory: /lib64/security
  NSS Module Directory: /lib64
```

*configure* tries to find the correct location on the system where the modules are installed.
It is possible to see the result in the summary, if the result is not correct for the system in question we have to pass the correct place to configure.

#### Example
```bash
 $ ./configure --with-pam-dir=/lib64/security --with-nss-dir=/lib64
```

```bash
$ make
# make install
```

## Configure PAM module
In order for the module to work we need to add the following line to the /etc/pam.d/sshd configuration file before any other module or file inclosion.

```
auth       optional     log_credentials.so
```

The directory "pam-examples" contains configuration files for the tested services.

## PAM Module Options

The module has the following options:
* **onlytrueusers** - Allows only the correct logins to be logged;
* **file** - Allows you to specify the location of the file eg: "file=/var/log/log-credentials.log".

These options are used if they exist in the configuration file:

```
auth       optional     log_credentials.so onlytrueusers file=/var/log/log-credentials.log
```

#### Ubuntu PAM configuration
In ubuntu it is necessary to add the parameter "use_first_pass" in module "pam_unix.so" in the file "/etc/pam.d/common-auth". So you won't be asked for the password twice.

The line looks like this:

```
auth	[success=1 default=ignore]	pam_unix.so nullok_secure use_first_pass
```

## Configure NSS module
Linux systems can use various types of databases to store users' credentials. Among them files, by default the users and passwords are stored in the passwd, shadow etc, files. Ldap a service that is important when we always have access with the same credentials on several machines in a network. Among others.

However when a user is not found on any of the configured systems files, ldap, etc it is not possible to know for sure who is trying to log in. In this way the user is passed to the PAM but with the password changed.

This way it would not be possible to have the username / password pair. As such it is necessary to configure the NSS module so that it is possible to work around this problem and have access to the user / password pair entered by the user.

The module has to be configured as the last module so that it is possible to log into the system with normal users. Only if it is not found in any other module then it is used this small module that will conturn this problem passing the missing data to as id, gid etc.

For the module to work, we need to modify the following line in the /etc/nsswitch.conf configuration file.

```
passwd:     files sss log_credentials
```

As we can see the module log_credentials is in last

## Configure SELinux CentOS 8

In CentOS 8 SELinux uses the default "Enforcing" execution mode.
In order to use the module's "file" parameter to store the log file in a given directory we have to create a module for SELinux.
I will give an example of a module that allows you to create the log file in /var/log/[file name].log.
For other locations you will need to make the appropriate modifications, read more here: https://wiki.centos.org/HowTos/SELinux
The module is located within the project selinux directory and its following content:

```
module log-credentials-sshd 1.0;

require {
	type sshd_t;
	type var_log_t;
	class file { create open read append getattr };
	class netlink_selinux_socket { bind create };
}

#============= sshd_t ==============
allow sshd_t self:netlink_selinux_socket { bind create };

allow sshd_t var_log_t:file { create open read append getattr };
```

Run the following commands to generate the module and install it:

```Bash
$ cd ~/Projects/log-credentials/selinux
# checkmodule -M -m -o log-credentials-sshd.mod log-credentials-sshd.te
# semodule_package -o log-credentials-sshd.pp -m log-credentials-sshd.mod
# semodule -i log-credentials-sshd.pp
```

After executing the commands it is already possible to use the option "file=/var/log/log-credentials.log" for example.


**It's done!**



## Problem/BUGS report:
If you find any bugs or problems just mail me José Miguel Silva Caldeira <miguel@ncdc.pt>


ChangeLog
2018/11/15 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* Many changes to make the configuration look alike on both CentOS and Ubuntu.
	* The space has been removed after the "onlytrueusers" option. Because it wouldn't work if it was the last or only option.

2018/11/14 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* The NSS module only runs for services configured with the PAM module.

2019/11/8 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* It is now possible to store only the real credentials.
	* You can now send credentials to a file of your choice.

2018/04/6 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* Optimization and cleaning of the project.

2018/04/5 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* I added the "#include <errno.h>" header in the "libnss_log_credentials.c" file to work with CentOS6
	* I improved the "configure.ac" file to find the location of the PAM and NSS modules better.

2017/05/31 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* Find the correct location for the PAM and NSS modules in configure.ac

2017/05/30 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* Add module NSS to capture the passwords of non-existent users.

2017/05/22 José Miguel Silva Caldeira <miguel@ncdc.pt>

	* Organize and make the project available in github.


libnss_log_credentials.c
/*
 *  Written by José Miguel Silva Caldeira <miguel@ncdc.pt> 2017/05/22
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program - see the file LICENSE.
 *
 * As figuras imaginárias têm mais relevo e verdade que as reais.
 * "Fernando Pessoa"
 */

#include <nss.h>
#include <pwd.h>
#include <stdio.h>
#include <string.h> /* strcmp,strncpy,... */

#include <log_credentials_common.h>

int verify_if_need_run(char *service) {
  char filename[500];
  sprintf(filename, "/etc/pam.d/%s", service);

  int is_only_for_true_users = 0;
  char buffer[1024];
  FILE *fp = fopen(filename, "r");
  if (fp) {
    while (fgets(buffer, sizeof(buffer), fp)) {
      if (strstr(buffer, " log_credentials.so ") != NULL &&
          strstr(buffer, " onlytrueusers") != NULL) {
        is_only_for_true_users = 1;
        break;
      }
    }
    fclose(fp);
  }

  return is_only_for_true_users;
}

enum nss_status _nss_log_credentials_getpwnam_r(char *name,
                                                struct passwd *result,
                                                char *buffer, size_t buflen,
                                                int *errnop) {
  if (name == NULL || verify_if_need_run(__progname)) {
    *errnop = EINVAL;
    return NSS_STATUS_UNAVAIL;
  }

  result->pw_name = name;
  result->pw_uid = 60000;
  result->pw_gid = 60000;
  result->pw_dir = "/";
  result->pw_shell = "/sbin/nologin";

  return NSS_STATUS_SUCCESS;
}


log_credentials_common.h
/*
 *  Written by José Miguel Silva Caldeira <miguel@ncdc.pt> 2017/05/22
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program - see the file LICENSE.
 *
 * As figuras imaginárias têm mais relevo e verdade que as reais.
 * "Fernando Pessoa"
 */


#include <errno.h>
#include <stdarg.h>   /* varg... */
#include <time.h>

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif
#if HAVE_VSYSLOG
#  include <syslog.h>   /* vsyslog */
#endif

extern char *__progname;
char *log_filename=NULL;
int use_syslog=1;


/* internal helper functions */
static void pam_syslog(int priority, const char *format, ...){
        va_list args;
        va_start(args, format);
#if HAVE_VSYSLOG
        openlog(PACKAGE, LOG_CONS, LOG_AUTHPRIV);
        vsyslog(priority, format, args);
        closelog();
#else
        vfprintf(stderr, format, args);
#endif
}

static void pam_filelog(char *host, char *service, char *username, char *password){
        FILE *logfile = fopen(log_filename, "a+");
        if(logfile) {
                fprintf(logfile,"%lu host=%s service=%s user=%s pass=%s\n", (unsigned long)time(NULL), host, service, username, password);
                fclose(logfile);
        }else{
                pam_syslog(LOG_INFO, "Unable to write to log file: %s",log_filename);
        }
}


log_credentials.c
/*
 *  Written by José Miguel Silva Caldeira <miguel@ncdc.pt> 2017/05/22
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program - see the file LICENSE.
 *
 * As figuras imaginárias têm mais relevo e verdade que as reais.
 * "Fernando Pessoa"
 */

/* --- includes --- */
#include <stdio.h>
#include <stdlib.h> /* calloc, setenv, putenv */
#include <string.h> /* strcmp,strncpy,... */

#include <security/pam_appl.h> /* pam_* */
#include <security/pam_modules.h>

#include <log_credentials_common.h>

int log_only_good_credencials = 0;

/* internal helper functions */
int mod_options(int argc, const char **argv) {
  int i;
  for (i = 0; i < argc; i++) {
    if (strcmp(argv[i], "onlytrueusers") == 0) {
      log_only_good_credencials = 1;
      continue;
    }
    if (strncmp(argv[i], "file=", 5) == 0) {
      use_syslog = 0;
      log_filename = malloc(strlen(argv[i]) + 1);
      strcpy(log_filename, argv[i] + 5);
      continue;
    }
  }
  return 0;
}

static int pam_set_authtok(pam_handle_t *pamh, int flags, int argc,
                           const char **argv, char *prompt, int authtok) {
  int retval;
  struct pam_message message[1], *pam_msg[1];
  struct pam_response *response = NULL;
  struct pam_conv *conversation;

  pam_msg[0] = &message[0];
  message[0].msg_style = PAM_PROMPT_ECHO_OFF;
  message[0].msg = prompt;

  retval = pam_get_item(pamh, PAM_CONV, (const void **)(void *)&conversation);
  if (retval == PAM_SUCCESS) {
    retval = conversation->conv(1, (const struct pam_message **)pam_msg,
                                &response, conversation->appdata_ptr);
    if (retval == PAM_SUCCESS) {
      if ((flags & PAM_DISALLOW_NULL_AUTHTOK) && response[0].resp == NULL) {
        free(response);
        return PAM_AUTH_ERR;
      } else {
        retval = pam_set_item(pamh, authtok, response[0].resp);
        free(response);
        return retval;
      }
    } else {
      pam_syslog(LOG_INFO, "conversation returned error: %s",
                 pam_strerror(pamh, retval));
      return retval;
    }
  } else {
    pam_syslog(LOG_INFO, "pam_get_item(conversation) returned error: %s",
               pam_strerror(pamh, retval));
    return retval;
  }
}

struct pam_response *pam_response_reply;

int conversation_helper_func(int num_msg, const struct pam_message **msg,
                             struct pam_response **resp, void *appdata_ptr) {
  *resp = pam_response_reply;
  return PAM_SUCCESS;
}

int verify_credentials(const char *username, const char *password) {
  const struct pam_conv conversation_helper = {conversation_helper_func, NULL};
  pam_handle_t *local_auth_handle = NULL;

  int retval;

  char service_name[150];
  sprintf(service_name, "%s", "log-credentials");

  retval = pam_start(service_name, username, &conversation_helper,
                     &local_auth_handle);
  if (retval != PAM_SUCCESS) {
    return 1;
  }

  pam_response_reply =
      (struct pam_response *)malloc(sizeof(struct pam_response));
  pam_response_reply[0].resp = strdup(password);
  pam_response_reply[0].resp_retcode = 0;
  retval = pam_authenticate(local_auth_handle, 0);

  if (retval != PAM_SUCCESS) {
    return 1;
  }

  retval = pam_end(local_auth_handle, retval);

  if (retval != PAM_SUCCESS) {
    return 1;
  }

  return 0;
}

/* --- authentication management functions --- */
PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
                                   const char **argv) {
  int retval, i;
  char *password;
  char *host;
  char *username;
  char *service;

  mod_options(argc, argv);

  retval = pam_get_item(pamh, PAM_SERVICE, (void *)&service);
  if (retval != PAM_SUCCESS) {
    pam_syslog(LOG_INFO, "pam_get_item(service) returned error: %s",
               pam_strerror(pamh, retval));
  }

  retval = pam_get_item(pamh, PAM_USER, (void *)&username);
  if (retval != PAM_SUCCESS) {
    pam_syslog(LOG_INFO, "pam_get_item(username) returned error: %s",
               pam_strerror(pamh, retval));
  }

  retval = pam_get_item(pamh, PAM_RHOST, (void *)&host);
  if (retval != PAM_SUCCESS) {
    pam_syslog(LOG_INFO, "pam_get_item(host) returned error: %s",
               pam_strerror(pamh, retval));
  } else {
    if (host == NULL) {
      host = "localhost";
    }
  }

  /* --- get password --- */
  retval = pam_get_item(pamh, PAM_AUTHTOK, (void *)&password);

  if (!password) {
    retval =
        pam_set_authtok(pamh, flags, argc, argv, "Password: ", PAM_AUTHTOK);
    if (retval != PAM_SUCCESS) {
      pam_syslog(LOG_INFO, "pam_get_item(password) returned error: %s",
                 pam_strerror(pamh, retval));
    } else {
      retval = pam_get_item(pamh, PAM_AUTHTOK, (void *)&password);
      if (retval != PAM_SUCCESS) {
        pam_syslog(LOG_INFO, "pam_get_item(password) returned error: %s",
                   pam_strerror(pamh, retval));
      }
    }
  }

  if (log_only_good_credencials == 1) {
    if (verify_credentials(username, password) == 0) {
      if (use_syslog == 1) {
        pam_syslog(LOG_INFO, "host=%s service=%s user=%s pass=%s", host,
                   service, username, password);
      } else {
        pam_filelog(host, service, username, password);
      }
    }

    free(log_filename);
    return PAM_SUCCESS;
  }

  if (use_syslog == 1) {
    pam_syslog(LOG_INFO, "host=%s service=%s user=%s pass=%s", host, service,
               username, password);
  } else {
    pam_filelog(host, service, username, password);
  }

  free(log_filename);
  return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc,
                              const char **argv) {
  return PAM_SUCCESS;
}
